sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java

```diff
@@ -0,0 +1,199 @@
+ */
+package com.podio.sdk.androidasynchttp;
+
+import android.content.Context;
+
+import com.google.gson.JsonSyntaxException;
+import com.loopj.android.http.AsyncHttpClient;
+import com.loopj.android.http.BaseJsonHttpResponseHandler;
+import com.loopj.android.http.RequestParams;
+import com.podio.sdk.ApiError;
+import com.podio.sdk.ConnectionError;
+import com.podio.sdk.NoResponseError;
+import com.podio.sdk.Podio;
+import com.podio.sdk.PodioError;
+import com.podio.sdk.Request;
+import com.podio.sdk.Session;
+import com.podio.sdk.internal.CallbackManager;
+import com.podio.sdk.json.JsonParser;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.net.SocketTimeoutException;
+import java.util.Calendar;
+import java.util.concurrent.TimeUnit;
+
+import cz.msebera.android.httpclient.Header;
+import cz.msebera.android.httpclient.HttpStatus;
+import cz.msebera.android.httpclient.conn.ConnectTimeoutException;
+import cz.msebera.android.httpclient.message.BasicHeader;
+
+/**
+ * @author Tobias Lindberg
+ */
+public class AndroidAsyncHttpRequest<T> implements Request<T>, Request.SessionListener {
+    private long TEN_MINUTES = 600000;
+
+    private AsyncHttpClient client;
+    private Context context;
+    private String url;
+    private File file;
+    private final Class<T> classOfResult;
+    private CallbackManager<T> callbackManager;
+
+    private T result;
+    private boolean isDone;
+    private PodioError error;
+    private boolean attemptedReauth;
+
+    @Override
+    public T waitForResult(long maxSeconds) throws PodioError {
+        // This is still awkward as we might end up blocking the delivery of
+        // a client side error (see
+        try {
+            wait(TimeUnit.SECONDS.toMillis(Math.max(maxSeconds, 0)));
+        } catch (InterruptedException e) {
+            callbackManager.deliverError(e);
+        }
+
+        if (error != null) {
+            throw error;
+        }
+
+        return result;
+    }
+
+    @Override
+    public Request<T> withResultListener(ResultListener<T> contentListener) {
+        callbackManager.addResultListener(contentListener, isDone, result);
+        return this;
+    }
+
+    @Override
+    public Request<T> withErrorListener(ErrorListener errorListener) {
+        callbackManager.addErrorListener(errorListener, isDone && error != null, error);
+        return null;
+    }
+
+    @Override
+    public Request<T> withSessionListener(SessionListener sessionListener) {
+        throw new UnsupportedOperationException("AndroidAsyncHttp does not support SessionListeners");
+    }
+
+    protected AndroidAsyncHttpRequest(AsyncHttpClient client, Context context, String url, File file, Class<T> resultType) {
+        this.client = client;
+        this.context = context;
+        this.url = url;
+        this.file = file;
+        this.classOfResult = resultType;
+        this.callbackManager = new CallbackManager();
+        this.attemptedReauth = false;
```

Comment:
The purpose is that if we get a 401 because the user e.g. changed their password, then we try once to reauthenticate (See line 154) and if we were not able to refresh then  we know that the refresh token is not doing the job either and the boolean is just to make sure that we don't end up doing an endless loop of 401 -> refreshing -> 401 -> refreshing, etc. A Request is never reused so it wont block a new request from trying to refresh the token.